lib/ide/src/controller.rs

```diff
@@ -0,0 +1,426 @@
+    ///
+    /// The remote connections should be already established.
+    pub fn new(file_manager_transport:impl Transport + 'static) -> StrongHandle {
+        let file_manager = file_manager_client::Client::new(file_manager_transport);
+        let ret = Data {
+            file_manager,
+            module_cache: default(),
+        };
+        StrongHandle::new(ret)
+    }
+
+    /// Project controller's state.
+    #[derive(Debug)]
+    pub struct Data {
+        /// File Manager Client.
+        file_manager : file_manager_client::Client,
+        /// Cache of module controllers. As the weak is handle (we don't manage
+        /// their lifetime), dead ones might still be included.
+        module_cache : HashMap<module::Location, module::WeakHandle>,
+    }
+
+    impl Data {
+        /// Returns handle to given module's controller if already present.
+        pub fn lookup_module(&mut self, loc:&module::Location) -> Option<module::StrongHandle> {
+            let weak = self.module_cache.get(loc)?;
+            match weak.upgrade() {
+                Some(strong) => Some(strong),
+                None => {
+                    self.module_cache.remove(loc);
+                    None
+                }
+            }
+        }
+
+        /// Stores given module controller handle and returns it.
+        ///
+        /// Note: handle stored in the project controller is weak, so the caller
+        /// remains responsible for managing the module's controller lifetime
+        /// (this is why the strong handle is returned).
+        ///
+        /// If there was already another module controller present in the cache,
+        /// it will be overwritten. Typically caller should use `lookup_module`
+        /// first.
+        pub fn insert_module(&mut self, data:module::Data) -> module::StrongHandle {
+            let path        = data.loc.clone();
+            let module      = module::StrongHandle::new(data);
+            let module_weak = module.downgrade();
+            self.module_cache.insert(path, module_weak);
+            module
+        }
+
+        /// Obtains a handle to a module controller interested in this
+        /// filesystem event.
+        pub fn relevant_module
+        (&mut self, event:&file_manager_client::Event) -> Option<module::StrongHandle> {
+            let location = Self::relevant_location(event)?;
+            self.lookup_module(&location)
+        }
+
+        /// Identifies module affected by given file manager's event.
+        pub fn relevant_location(event:&file_manager_client::Event) -> Option<module::Location> {
+            use file_manager_client::Event;
+            use file_manager_client::Notification::FilesystemEvent;
+            match event {
+                Event::Closed          => None,
+                Event::Error(_)        => None,
+                Event::Notification(n) => match n {
+                    FilesystemEvent(e) => Some(module::Location(e.path.0.clone()))
```

Comment:
Is this pricy or cheap clone? If the later, please use `clone_ref` instead.

Reply:
fixed on my branch `wip/ao/text-controller`